File: jar/src/main/java/jline/api/mc/Ctmc_solve_reducible.java
package jline.api.mc;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import jline.util.Pair;
import jline.util.matrix.Matrix;

public final class Ctmc_solve_reducible {
    private Ctmc_solve_reducible() {}

    /**
     * Result of full reducible CTMC solve.
     */
    public static final class CtmcSolveReducibleResult {
        public final Matrix pi;
        public final List<Matrix> pis;
        public final Matrix pi0;
        public final List<List<Integer>> scc;
        public final List<Boolean> isrec;

        public CtmcSolveReducibleResult(Matrix pi, List<Matrix> pis, Matrix pi0,
                                        List<List<Integer>> scc, List<Boolean> isrec) {
            this.pi = pi;
            this.pis = pis;
            this.pi0 = pi0;
            this.scc = scc;
            this.isrec = isrec;
        }
    }

    private static Map<String, Object> defaultOptions() {
        Map<String, Object> opts = new HashMap<String, Object>();
        opts.put("tol", 1e-12);
        return opts;
    }

    /**
     * Solve reducible CTMCs by converting to DTMC via randomization.
     *
     * @param Q Infinitesimal generator matrix (possibly reducible)
     * @param pi0 Initial state distribution (may be null)
     * @param options Solution options
     * @return Pair of (pi: steady-state distribution, scc: strongly connected components info)
     */
    public static Pair<Matrix, List<List<Integer>>> ctmc_solve_reducible(
            Matrix Q, Matrix pi0, Map<String, Object> options) {
        // Convert CTMC to DTMC via randomization
        Pair<Matrix, Double> rand = Ctmc_randomization.ctmc_randomization(Q);
        Matrix P = rand.getLeft();

        // Solve the reducible DTMC
        Pair<Matrix, List<List<Integer>>> dtmcResult =
                Dtmc_solve_reducible.dtmc_solve_reducible(P, pi0, options);
        Matrix pi = dtmcResult.getLeft();
        List<List<Integer>> scc = dtmcResult.getRight();

        return new Pair<Matrix, List<List<Integer>>>(pi, scc);
    }

    public static Pair<Matrix, List<List<Integer>>> ctmc_solve_reducible(Matrix Q) {
        return ctmc_solve_reducible(Q, null, defaultOptions());
    }

    public static Pair<Matrix, List<List<Integer>>> ctmc_solve_reducible(Matrix Q, Matrix pi0) {
        return ctmc_solve_reducible(Q, pi0, defaultOptions());
    }

    /**
     * Alternative signature that returns additional information.
     */
    public static CtmcSolveReducibleResult ctmc_solve_reducible_full(
            Matrix Q, Matrix pi0, Map<String, Object> options) {
        // Convert CTMC to DTMC via randomization
        Pair<Matrix, Double> rand = Ctmc_randomization.ctmc_randomization(Q);
        Matrix P = rand.getLeft();

        // Solve the reducible DTMC (full implementation)
        Dtmc_solve_reducible.DtmcSolveReducibleResult result =
                Dtmc_solve_reducible.dtmc_solve_reducible_full(P, pi0, options);

        return new CtmcSolveReducibleResult(result.pi, result.pis, result.pi0, result.scc, result.isrec);
    }

    public static CtmcSolveReducibleResult ctmc_solve_reducible_full(Matrix Q) {
        return ctmc_solve_reducible_full(Q, null, defaultOptions());
    }

    public static CtmcSolveReducibleResult ctmc_solve_reducible_full(Matrix Q, Matrix pi0) {
        return ctmc_solve_reducible_full(Q, pi0, defaultOptions());
    }
}

File: jar/src/main/kotlin/jline/api/mc/Ctmc_solve_reducible.kt
package jline.api.mc

import jline.util.matrix.Matrix

/**
 * Solve reducible CTMCs by converting to DTMC via randomization
 *
 * @param Q Infinitesimal generator matrix (possibly reducible)
 * @param pi0 Initial state distribution (optional)
 * @param options Solution options (optional)
 * @return Pair of (pi: steady-state distribution, scc: strongly connected components info)
 */
fun ctmc_solve_reducible(
    Q: Matrix,
    pi0: Matrix? = null,
    options: Map<String, Any> = mapOf("tol" to 1e-12)
): Pair<Matrix, List<List<Int>>> {
    // Convert CTMC to DTMC via randomization
    val (P, _) = ctmc_randomization(Q)

    // Solve the reducible DTMC
    val dtmcResult = dtmc_solve_reducible(P, pi0, options)
    val pi = dtmcResult.first
    val scc = dtmcResult.second

    return Pair(pi, scc)
}

/**
 * Alternative signature that returns additional information
 */
data class CtmcSolveReducibleResult(
    val pi: Matrix,
    val pis: List<Matrix>,
    val pi0: Matrix,
    val scc: List<List<Int>>,
    val isrec: List<Boolean>
)

fun ctmc_solve_reducible_full(
    Q: Matrix,
    pi0: Matrix? = null,
    options: Map<String, Any> = mapOf("tol" to 1e-12)
): CtmcSolveReducibleResult {
    // Convert CTMC to DTMC via randomization
    val (P, _) = ctmc_randomization(Q)

    // Solve the reducible DTMC (would need full implementation)
    val result = dtmc_solve_reducible_full(P, pi0, options)

    return CtmcSolveReducibleResult(
        pi = result.pi,
        pis = result.pis,
        pi0 = result.pi0,
        scc = result.scc,
        isrec = result.isrec
    )
}
/**
 * CTMC solve reducible algorithms
 */
@Suppress("unused")
class CtmcSolveReducibleAlgo {
    companion object {
        // Class documentation marker for Dokka
    }
}
